onnxruntime/core/common/threadpool.cc

```diff
@@ -26,6 +26,101 @@ namespace onnxruntime {
 
 namespace concurrency {
 
+void ThreadPoolProfiler::Start() {
+  std::lock_guard<OrtMutex> lock(mutex_);
+  auto per_thread_number = per_thread_numbers_.find(std::this_thread::get_id());
+  if (per_thread_number == per_thread_numbers_.end()) {
+    per_thread_numbers_.insert({std::this_thread::get_id(), {}});
+  }
+}
+
+std::string ThreadPoolProfiler::Stop() {
+  std::lock_guard<OrtMutex> lock(mutex_);
+  auto per_thread_number = per_thread_numbers_.find(std::this_thread::get_id());
+  if (per_thread_number == per_thread_numbers_.end()) {
+    return {};
+  } else {
+    return per_thread_number->second.Reset();
+  }
+}
+
+void ThreadPoolProfiler::LogStart() {
+  if (per_thread_numbers_.empty()) {
```

Comment:
>(per_thread_numbers_.empty()) [](start = 5, length = 29)

i think this check would need to be guarded by mutex_ as well, since other threads may possibly be writing per_thread_numbers_.
is the performance overhead from locking mutex_ to access the per thread measurements acceptable? do we expect much contention?

Reply:
we could afford to miss a few numbers which supposed to be collected.
What we cannot afford is the overhead of mutexing, and yes the contention would be very much heated.
Still looking for ways to replace mutexing.